### PR TITLE
Add non-normative DID Resolution and Dereferencing sections to Architecture.

### PR DESCRIPTION
**Pull request by msporny**
These are the non-normative changes from @jricher's resolution contract PR. Moving them out to make it easier to merge the bits everyone seems to agree on. 

I've done a number of editorial fixes to bring it more in line with the way the DNS and URI specs talk about resolvers and dereferencing, align the language with the plan for the architecture section, move definitions of terms into the terminology section, and various spurious whitespace and margin overrun issues.

In theory, if you were fine with #253, you should be fine with this one as well as it's just editorial changes to #253.

PRs #263, #264, and #265 layer the normative language on top of this PR in more easily digestible section-based chunks.

/cc @jricher (since he's still not in the Github w3c wg list)

***
<a href="https://pr-preview.s3.amazonaws.com/w3c/did-core/pull/262.html" title="Last updated on May 1, 2020, 3:06 PM UTC (e83669a)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/w3c/did-core/262/cfc07f9...e83669a.html" title="Last updated on May 1, 2020, 3:06 PM UTC (e83669a)">Diff</a>

### CODE REVIEW

File: index.html [hunk 1]
```diff
@@ -1076,11 +1113,11 @@ <h2>Fragment</h2>
   document</a> to identify a component of the document (for example, a unique
   <a>public key description</a> or <a>service endpoint</a>). <a>DID fragment</a>
   syntax and semantics are identical to a generic URI fragment and  MUST conform
-  to <a data-cite="RFC3986#section-3.5">RFC&nbsp;3986, section 3.5</a>. To resolve
+  to <a data-cite="RFC3986#section-3.5">RFC&nbsp;3986, section 3.5</a>. To dereference
   a <a>DID fragment</a> reference, the complete <a>DID URL</a> including the
-  <a>DID fragment</a> MUST be used as input to the <a>DID URL</a> dereferencing
-  algorithm for the target component in the <a>DID document</a> object. For more
-  information, see [[?DID-RESOLUTION]].
+  <a>DID fragment</a> MUST be used as input to the <a>DID URL dereferencing</a>
```
Comment (kdenhartog) [inline index.html:1118]: just catching this now. Would it be considered derefencing if I used `did:example:123?id=key1`? I know we still have to discuss the usage of id query parameter, but figured it should be considered here as well.

File: index.html [hunk 2]
```diff
@@ -1991,7 +2028,8 @@ <h1>Core Representations</h1>
     <p>
 Producers MUST indicate which representation of a document has been used via a media type
 in the document's metadata. Consumers MUST determine which representation a document is in
-via a media type in the document's metadata. Consumers MUST NOT determine the representation
+via the <code>content-type</code> DID resolver metadata field. (See <a href="#did-resolution"></a>).
+Consumers MUST NOT determine the representation
```
Comment (kdenhartog) [inline index.html:2032]: Any particular reason to use MUST NOT here rather than "Consumers MUST determine the representation using the content-type and content."?

**Comment by iherman** [comment]
(Absolutely aside:

> /cc @jricher (since he's still not in the Github w3c wg list)

He is now. No idea why this wasn't done before)

File: index.html [hunk 3]
```diff
@@ -2019,8 +2057,8 @@ <h2>
         </h2>
 
         <p>
-When producing and consuming DID documents that are in plain JSON (as noted by
-the document metadata), the following rules MUST be followed.
+When producing and consuming DID documents that are in plain JSON (as indicated by
+a <code>content-type</code> of <code>application/json+did</code> in the resolver metadata), the following rules MUST be followed.
```
Comment (iherman) [inline index.html:2061]: <code>application/json+did</code> does not sound right. Shouldn't it be <code>application/did+json</code>?

File: index.html [hunk 4]
```diff
@@ -2161,8 +2199,8 @@ <h2>
         </p>
 
         <p>
-When producing and consuming DID documents that are in JSON-LD (as noted by
-the document metadata), the following rules MUST be followed.
+When producing and consuming DID documents that are in JSON-LD (as indicated by
+a <code>content-type</code> of <code>application/jsonld+did</code> in the resolver metadata), the following rules MUST be followed.
         </p>
```
Comment (iherman) [inline index.html:2204]: Same remark as above: <code>application/json+did</code> does not sound right. Shouldn't it be <code>application/did+json</code>?

**Comment by iherman** [comment]
Just something weird... When I push on the 'preview' button, then what I get is _not_ what I seem to see in the 'changes'. And the header above says:

> msporny wants to merge 4 commits into master from jricher-resolution-nonnormative

are we sure this is a PR on the text of Justin, and a possible merge would not merge into the master? I am not sure that is what we want, do we?

Cc @msporny

**Comment by jricher** [comment]
-1

This ignores the most important part of #253 in an attempt to sideline the discussion there. Please close this pull request and add comments there.

File: index.html [hunk 5]
```diff
@@ -630,16 +630,21 @@ <h2>DID Documents</h2>
...
-A <a>DID resolver</a> is a software and/or hardware component that is capable of
-executing the <a>DID resolution</a> function for at least one
-<a>DID method</a>.
+A <a>DID resolver</a> is a software and/or hardware component that takes
+a <a>DID</a> as input and produces a <a>conforming DID document</a> as output.
+This process is called <a>DID resolution</a> and often includes
+resolution input options and resolution output metadata. The general process of
+DID Resolution is described in <a href="#did-resolution"></a>. The details of
+implementing a <a>DID resolver</a> is described in the DID Resolution
+specification [[?DID-RESOLUTION]].
```
Comment (jricher) [inline index.html]: This pushes off all details to another specification without providing anything actionable by implementors. I believe that this is intentional, but the language is dangerous and slippery.

Comment (jricher) [reply index.html]: Also note that this is much more than an "editorial change" to #253, as claimed by the header of the pull request. This one statement drastically changes the nature and intent of the pull request.

File: index.html (hunk 3 above)
Comment (jricher) [reply index.html:2061]: That's part of the discussion under #255

File: index.html (hunk 2 above)
Comment (jricher) [reply index.html:2032]: This was from a previous pull request #197 that I wrote so I can address it -- what we were trying to do is say that you can't just throw the parser at something to see if it fails or not. So for example you can't get something, throw it in your JSON parser, and then look for the context field to figure out if it's JSON LD. By this definition, it's not JSON LD unless it's got the appropriate JSON LD content type associated with it.

Comment (kdenhartog) [reply index.html:2032]: Makes sense, thanks for the clarification. I was a bit confused about how to test this, but now that you point out that case it makes sense that the test would need to check that the JSON-LD aspects are valid.

File: index.html (hunk 1 above)
Comment (msporny) [reply index.html:1118]: > I noticed the contract has been removed at this point, but I'm hoping it will come back with your addition of normative statements.

Yes, that's planned in the next PR that layers on top of this one.

Comment (msporny) [reply index.html:1118]: > Would it be considered derefencing if I used did:example:123?id=key1?

Yes, given that id=key1 becomes a DID Parameter whose purpose is to dereference to a key in the DID Document.

That said, I think that doing this is a bad idea. :)

File: index.html (hunk 3 above)
Comment (msporny) [reply index.html:2061]: Yes, this is wrong per RFC6838 and should be fixed. Thanks for catching that @iherman.

The specific section is here:

https://tools.ietf.org/html/rfc6838#section-4.2.8

More elaboration here:

https://tools.ietf.org/html/rfc6839

Comment (msporny) [reply index.html]: Fixed in 4f1c79ce416dc2564e693d76cd63ede0242ccd80.

File: index.html (hunk 4 above)
Comment (msporny) [reply index.html:2206]: Correct, that was wrong, thanks @iherman!

Fixed in 4f1c79ce416dc2564e693d76cd63ede0242ccd80.

**Comment by msporny** [comment]
> > msporny wants to merge 4 commits into master from jricher-resolution-nonnormative
> 
> are we sure this is a PR on the text of Justin, and a possible merge would not merge into the master? I am not sure that is what we want, do we?

If I wanted to do that, I'd have to fork Justin's fork, which is a fork of this repo. It's a more difficult than necessary way to merge the changes in. 

What I did instead was to pull @jricher's PR directly into a branch in the WG repo (so all Editors could have access to the text), squashed his edits (removing unnecessary middle-edits and the normative bits that will be added in a separate PR), and then performed editorial fixes to the rest. I did a line-by-line comparison to his non-normative text in https://github.com/w3c/did-core/pull/262/commits/cee116acf2359e01c2236f460302289e7e6319aa and then did my changes on top of that.

This will just let us merge the non-normative stuff, w/ Editor corrections, when we have a clean review on this PR rather than having to completely agree to the normative and non-normative parts all at the same time (which will take much longer than getting agreement on the non-normative portions and then the normative portions before we can merge anything).

**Comment by msporny** [comment]
> This ignores the most important part of #253 in an attempt to sideline the discussion there. 

This PR pulls the non-normative portions out of #253, performs a number of editorial fixes on it and prepares it for merging into the master branch of the specification.

The normative portions of #253 are coming in a PR shortly after this one, as stated in the text opening this PR -- https://github.com/w3c/did-core/pull/262#issue-405772664:

> The next PR is going to layer the normative language in #253 on top of this PR.

It is not an attempt to sideline discussion, it is an attempt to split #253 into more easily digestible chunks so that it can be merged into the specification more rapidly.

File: index.html (hunk 5 above)
Comment (msporny) [reply index.html]: See https://github.com/w3c/did-core/pull/262#issuecomment-616893036.

File: index.html [hunk 6]
```diff
@@ -630,7 +630,43 @@ <h2>DID Documents</h2>
     </section>
 
     <section>
-      <h2>DID Resolvers and DID Resolution</h2>
+      <h2>DID Resolvers</h2>
+
+      <p>
+A <a>DID resolver</a> is a software and/or hardware component that takes a
```
Comment (jricher) [inline index.html:636]: It's confusing to talk about the component first and then the process that defines them, I would suggest that this be reversed. However, realizing that this section is defining components of the architecture and not actions of the architecture, this ordering might be desired. With that in mind, perhaps these pieces should be moved to the resolution section below instead.

File: index.html [hunk 7]
```diff
@@ -630,7 +630,43 @@ <h2>DID Documents</h2>
     </section>
 
     <section>
-      <h2>DID Resolvers and DID Resolution</h2>
+      <h2>DID Resolvers</h2>
```
[the remark this answers is not in the document]
Comment (jricher) [reply index.html:633]: Or better yet, remove this section and incorporate its content into the resolution and dereferencing contract sections below.

File: index.html [hunk 8]
```diff
@@ -630,7 +630,43 @@ <h2>DID Documents</h2>
...
+of a <a>DID URL</a> to obtain a <a>DID document</a>. The <a>DID URL
+dereferencer</a> can then perform additional dereferencing on the <a>DID
+document</a> using the absolute <a>DID URL</a>.
+      </p>
+
+      <p>
+The general process of <a>DID URL dereferencing</a> is described in
+<a href="#did-url-dereferencing"></a>. The details of implementing a <a>DID URL
+dereferencer</a> are described in the DID Resolution specification
+[[?DID-RESOLUTION]].
```
Comment (jricher) [inline index.html:668]:
```suggestion
The inputs and outputs of the <a>DID URL dereferencing</a> process are defined in
<a href="#did-url-dereferencing"></a>. Additional considerations for implementing a <a>DID URL
dereferencer</a> are discussed in the DID Resolution specification
[[?DID-RESOLUTION]].
```

File: index.html [hunk 9]
```diff
@@ -630,7 +630,43 @@ <h2>DID Documents</h2>
...
+      <p>
+A <a>DID URL dereferencer</a> is a software and/or hardware component that takes
+a <a>DID URL</a> as input and produces a resource, which might or might not be a
+<a>conforming DID Document</a> as output. This process is called <a>DID URL
+dereferencing</a> and often includes dereferencing input options and
+dereferencing output metadata. Generally, a <a>DID URL dereferencer</a> uses a
+<a>DID resolver</a> to execute <a>DID resolution</a> on the <a>DID</a> portion
+of a <a>DID URL</a> to obtain a <a>DID document</a>. The <a>DID URL
+dereferencer</a> can then perform additional dereferencing on the <a>DID
+document</a> using the absolute <a>DID URL</a>.
```
Comment (jricher) [inline index.html:661]:
```suggestion
document</a> using the complete <a>DID URL</a>.
```

File: index.html [hunk 10]
```diff
@@ -630,7 +630,43 @@ <h2>DID Documents</h2>
...
+A <a>DID resolver</a> is a software and/or hardware component that takes a
+<a>DID</a> as input and produces a <a>conforming DID document</a> as output.
+This process is called <a>DID resolution</a> and often includes resolution input
+options and resolution output metadata.
+      </p>
+      <p>
+The general process of <a>DID resolution</a> is described in
+<a href="#did-resolution"></a>. The details of implementing a
+<a>DID resolver</a> are described in the DID Resolution specification
+[[?DID-RESOLUTION]].
```
Comment (jricher) [inline index.html:645]:
```suggestion
The inputs and outputs of the <a>DID resolution</a> process are defined in
<a href="#did-resolution"></a>. Additional considerations for implementing a
<a>DID resolver</a> are discussed in the DID Resolution specification
[[?DID-RESOLUTION]].
```

File: index.html [hunk 11]
```diff
@@ -1076,11 +1113,11 @@ <h2>Fragment</h2>
   document</a> to identify a component of the document (for example, a unique
   <a>public key description</a> or <a>service endpoint</a>). <a>DID fragment</a>
   syntax and semantics are identical to a generic URI fragment and  MUST conform
-  to <a data-cite="RFC3986#section-3.5">RFC&nbsp;3986, section 3.5</a>. To resolve
+  to <a data-cite="RFC3986#section-3.5">RFC&nbsp;3986, section 3.5</a>. To dereference
   a <a>DID fragment</a> reference, the complete <a>DID URL</a> including the
```
Comment (jricher) [inline index.html:1117]:
```suggestion
  a <a>DID fragment</a>, the complete <a>DID URL</a> including the
```

File: terms.html [hunk 12]
```diff
@@ -123,12 +130,22 @@
 verifiable data registry. For more information, see [[VC-DATA-MODEL]].
   </dd>
 
+  <dt><dfn>DID resolution</dfn></dt>
+
+  <dd>
+The function that takes as its input a <a>DID</a> and a set of resolution
+options and returns a <a>DID document</a> in a conforming representation
```
Comment (jricher) [inline terms.html:137]: Suggest shifting to use "input metadata" instead of "resolver options" here and throughout. WG should discuss actual terms.

```suggestion
The function that takes as its input a <a>DID</a> and a set of input
metadata and returns a <a>DID document</a> in a conforming representation
```

File: terms.html [hunk 13]
```diff
@@ -123,12 +130,22 @@
...
+plus additional metadata. This function relies on the "Read" operation of the
+applicable <a>DID method</a>. The contract of this function is defined
```
Comment (jricher) [inline terms.html:139]:
```suggestion
applicable <a>DID method</a>. The inputs and outputs of this function are defined
```

File: terms.html [hunk 14]
```diff
@@ -123,12 +130,22 @@
...
+
   <dt><dfn data-lt="DID resolvers">DID resolver</dfn></dt>
 
   <dd>
-A system that is capable of retrieving a <a>DID document</a> for a given
-<a>DID</a>. These systems are specified in the DID Resolution specification
-[[DID-RESOLUTION]].
+A <a>DID resolver</a> is a software and/or hardware component that takes
+a <a>DID</a> as input and produces a <a>conforming DID document</a> as
+output.
```
Comment (jricher) [inline terms.html:148]: Suggest we define the component in terms of the function it implements.

```suggestion
A <a>DID resolver</a> is a software and/or hardware component that takes
a <a>DID</a> as input and produces a <a>conforming DID document</a> as
output by performing <a>DID resolution</a>.
```

File: terms.html [hunk 15]
```diff
@@ -154,6 +171,27 @@
...
+  <dt><dfn>DID URL dereferencing</dfn></dt>
+
+  <dd>
+The function that takes as its input a <a>DID URL</a>, a <a>DID document</a>,
+and a set of dereferencing options and returns a resource, which might be
+a <a>DID document</a>, plus additional metadata. This function can use the <a>DID resolution</a>
+function to fetch a <a>DID document</a> indicated by the <a>DID</a> within the <a>DID URL</a>.
+The dereferencing function then can perform additional processing on the <a>DID document</a> to return
+the dereferenced resource indicated by the <a>DID URL</a>. The contract of this
+function is defined in <a href="#did-url-dereferencing"></a>.
```
Comment (jricher) [inline terms.html:183]:
```suggestion
the dereferenced resource indicated by the <a>DID URL</a>. The inputs and outputs of this
function are defined in <a href="#did-url-dereferencing"></a>.
```

File: terms.html [hunk 16]
```diff
@@ -154,6 +171,27 @@
...
+  </dd>
+
+  <dt><dfn data-lt"DID URL dereferencers">DID URL dereferencer</dfn></dt>
+
+  <dd>
+A software and/or hardware system that is capable of executing the <a>DID URL dereferencing</a>
+function, retrieving some form of data, which might or might not be a <a>DID document</a>) for a given
+<a>DID URL</a>. The details of systems are specified in the DID Resolution specification
+[[?DID-RESOLUTION]].
```
Comment (jricher) [inline terms.html:192]: Suggest we remove the outbound reference here and define component in terms of the function it implements.

```suggestion
A software and/or hardware system that is capable of executing the <a>DID URL dereferencing</a>
function, retrieving some form of data, which might or might not be a <a>DID document</a>) for a given
<a>DID URL</a> by performing <a>DID URL dereferencing</a>.
```

File: index.html [hunk 17]
```diff
@@ -2544,17 +2584,29 @@ <h1>
...
+
+    <section>
+        <h2>
+DID URL Dereferencing
+        </h2>
+    </section>
+
+    <section>
+        <h2>
+Resolution Metadata
```
Comment (jricher) [inline index.html:2607]: This is not an accurate name for the section as it covers both input and output metadata for both resolution and dereferencing. Suggest keeping the proposed name while other pull requests are still in progress so as not to lose synchronization.

```suggestion
Metadata Structure
```

File: index.html [hunk 18]
```diff
@@ -154,18 +154,18 @@
...
+of a <a>DID</a> to prove control over it and to be implemented independently
+of any centralized registry, identity provider, or certificate authority.
+<a>DID</a>s are URLs that associate a <a>DID subject</a> with a
+<a>DID document</a> allowing trustable interactions associated with that subject.
+Each <a>DID document</a> can express cryptographic material, verification methods,
+or <a>service endpoints</a>, which provide a set of mechanisms enabling a
+<a>DID controller</a> to prove control of the <a>DID</a>. <a>Service
+endpoints</a> enable trusted interactions associated with the <a>DID
+subject</a>. A <a>DID document</a> might contain semantics about the subject
+that it identifies. A <a>DID document</a> might contain the <a>DID subject</a>
```
Comment (jricher) [inline index.html:168]: Do these need to be in this specific pull request? There are several similar changes here. Space cleanup commits like this should be left to separate pull requests so as not to distract from discussion on content.

File: index.html [hunk 19]
```diff
@@ -630,7 +630,43 @@ <h2>DID Documents</h2>
...
+<a>DID resolver</a> are described in the DID Resolution specification
+[[?DID-RESOLUTION]].
+      </p>
+    </section>
+
+    <section>
+      <h2>DID Dereferencers</h2>
+
+      <p>
+A <a>DID URL dereferencer</a> is a software and/or hardware component that takes
```
Comment (jricher) [inline index.html:653]: Same comment about ordering above: define the process, then the component that provides that process.

File: index.html [hunk 20]
```diff
@@ -630,7 +630,43 @@ <h2>DID Documents</h2>
...
 
     <section>
-      <h2>DID Resolvers and DID Resolution</h2>
+      <h2>DID Resolvers</h2>
+
+      <p>
+A <a>DID resolver</a> is a software and/or hardware component that takes a
+<a>DID</a> as input and produces a <a>conforming DID document</a> as output.
+This process is called <a>DID resolution</a> and often includes resolution input
+options and resolution output metadata.
```
Comment (jricher) [inline index.html:639]: Suggest include input and output metadata in the definition. Phrasing "often includes" implies much more variability than is intended here. 

```suggestion
<a>DID</a> (and associated options) as input and produces a <a>conforming DID document</a> (and associated metadata) as output in a process called <a>DID resolution</a>.
```

File: index.html [hunk 21]
```diff
@@ -630,7 +630,43 @@ <h2>DID Documents</h2>
...
+
+    <section>
+      <h2>DID Dereferencers</h2>
+
+      <p>
+A <a>DID URL dereferencer</a> is a software and/or hardware component that takes
+a <a>DID URL</a> as input and produces a resource, which might or might not be a
+<a>conforming DID Document</a> as output. This process is called <a>DID URL
+dereferencing</a> and often includes dereferencing input options and
+dereferencing output metadata. Generally, a <a>DID URL dereferencer</a> uses a
```
Comment (jricher) [inline index.html:657]: Same comment as above: Suggest include input and output metadata in the definition. Phrasing "often includes" implies much more variability than is intended here.

```suggestion
a <a>DID URL</a> as input (and associated options) and produces a resource (and
associated metadata) as output in a process called <a>DID URL
dereferencing</a>. The resource produced by <a>DID URL dereferencing</a> might or might not be a
<a>conforming DID Document</a>. Generally, a <a>DID URL dereferencer</a> uses a
```

File: index.html [hunk 22]
```diff
@@ -2544,17 +2584,29 @@ <h1>
     </h1>
 
     <p>
-A <a>DID resolver</a> is a software or hardware component with an API for
-resolving <a>DIDs</a> of at least one <a>DID method</a>. It executes the read
-operation for the <a>DID method</a> corresponding to the <a>DID</a> being
-resolved to obtain the authoritative <a>DID document</a>. For more information,
-see Section <a href="#read-verify"></a>.
+This section defines the inputs and outputs of the
+<a>DID resolution</a> and <a>DID URL dereferencing</a> functions. The
```
Comment (jricher) [inline index.html:2588]:
```suggestion
<a>DID resolution</a> and <a>DID URL dereferencing</a> processes. The
```

File: index.html (hunk 8 above)
Comment (jricher) [reply index.html:668]: @philarcher I am proposing to change the language back toward what I had in #253, is this a better framing of the connection between the documents for you?

Comment (philarcher) [reply index.html:668]: That's certainly better, yes, but the word specification is still a problem. [DID-RESOLUTION] is not a specification (I know it is, and IMO the decision not the put it on the rec track is self-harming and irrational but there we are). How about:

The inputs and outputs of the <a>DID URL dereferencing</a> process are defined in
<a href="#did-url-dereferencing"></a>. Additional considerations for implementing a <a>DID URLdereferencer</a>, that are out of scope for this document,  are discussed 
[[?DID-RESOLUTION]].

Could you live with that?

Comment (jricher) [reply index.html:668]: How about this, which I think lands a little better and is a bit more concise:

```suggestion
The inputs and outputs of the <a>DID URL dereferencing</a> process are defined in
<a href="#did-url-dereferencing"></a>. Additional considerations for implementing a <a>DID URL
dereferencer</a> are discussed in 
[[?DID-RESOLUTION]].
```

Comment (philarcher) [reply index.html:668]: Looks like we've met in the middle - thanks Justin - I'm happy with this.

File: index.html (hunk 10 above)
Comment (jricher) [reply index.html:645]:
```suggestion
The inputs and outputs of the <a>DID resolution</a> process are defined in
<a href="#did-resolution"></a>. Additional considerations for implementing a
<a>DID resolver</a> are discussed in [[?DID-RESOLUTION]].
```

**Comment by jricher** [comment]
@msporny I believe the currently suggested text changes address the concerns raised and can be incorporated. I'll make parallel changes to the other pull requests separately.

File: index.html (hunk 1 above)
Comment (kdenhartog) [reply index.html:1118]: > 
> 
> > Would it be considered derefencing if I used did:example:123?id=key1?
> 
> Yes, given that id=key1 becomes a DID Parameter whose purpose is to dereference to a key in the DID Document.
> 
> That said, I think that doing this is a bad idea. :)

Would doing that require an extended document loader to handle that de-referencing?

File: index.html (hunk 18 above)
Comment (msporny) [reply index.html:168]: Yes, agreed, ideally they would be. Merging anyway so that these are fixed when merged into the main document. Will rebase the other PRs so they show up there.

File: index.html (hunk 7 above)
[the remark this answers is not in the document]
Comment (msporny) [reply index.html:633]: As discussed on the call, leaving this as-is for now to talk about the "nouns" in the architecture, and then what the nouns do.

File: index.html (hunk 6 above)
Comment (msporny) [reply index.html:636]: As discussed on the call, leaving this as-is for now to talk about the "nouns" in the architecture, and then what the nouns do.

File: index.html (hunk 19 above)
Comment (msporny) [reply index.html:652]: As discussed on the call, leaving this as-is for now to talk about the "nouns" in the architecture, and then what the nouns do.

File: index.html (hunk 1 above)
Comment (msporny) [reply index.html:1116]: It's the concept of "get me this thing that is expressed in the DID Document using this syntax that duplicates the use of fragment identifiers" that is problematic. If you want the key, get the DID Document, look for the key's fragment identifier, and you're done... w/o needing a DID Parameter for that feature. It can be done w/ a simple convenience function.

**Comment by msporny** [comment]
@jricher, all suggestions have been merged. Given the positive reviews in #253, given that this PR is based on that one, and given that this is all non-normative, I'm going to merge this in the next 48 hours unless I hear objections. I'll double check to make sure this is okay to do w/ Chairs and Editors today.

**Comment by philarcher** [comment]
Looking at this afresh today, I remain content with the revised wording in the DID Resolvers and DID Dereferencers sections.

But...

I just spotted this in the section on Generic DID Parameters:

"The exact processing rules for these parameters are specified in [DID-RESOLUTION]."

Sorry to be awkward but, again, I believe it is not right to include a sentence like that in a normative section. Formally, nothing is _specified_ in DID-RESOLUTION. Reusing the language we've converged on in the other sections, how about changing that to

Additional considerations for processing these parameters are provided in [?DID-RESOLUTION]

**Comment by msporny** [comment]
> Sorry to be awkward but, again, I believe it is not right to include a sentence like that in a normative section. 

I have awkwardly applied your exact wording in https://github.com/w3c/did-core/pull/262/commits/758b6355a34d7db09a43987a4421978423a662a3.

File: index.html (hunk 1 above)
Comment (kdenhartog) [reply index.html:1116]: could `did:example:123?id=key1` be used as an iri as well if the producer chose to do so? Really just trying to understand the edge cases here. I stick to fragments here normally cause they're easier to deal with and commonly used by everyone.

File: index.html [hunk 23]
```diff
@@ -933,23 +967,24 @@ <h2>Generic DID URL Parameters</h2>
         </table>
 
         <p>
-  The exact processing rules for these parameters are specified in
-  [[?DID-RESOLUTION]].
+Additional considerations for processing these parameters are provided
+in [?DID-RESOLUTION].
```
Comment (jricher) [inline index.html:971]: Suggest this change to align with other mentions. This line wasn't part of the original pull request, but we should change it now to keep it consistent. 

```suggestion
Additional considerations for processing these parameters are discussed
in [?DID-RESOLUTION].
```
